### ISSUE TITLE
Package as JAR to run the toolchain independently

### ISSUE THREAD

**Issue by vhschlenker** [Mari-Wie/ArchCNL#45]
Packaging ArchCNL as a JAR would reduce the setup necessary to use it.
Issues like #40 would be getting much easier by "just" calling a JAR in the pipeline.

**Comment by vhschlenker** [comment]
The current status is the commit 2885675fa376830f6a405eeb7f3f266e9657b729.
By calling `mvn package` a jar is produced which can be executed with `java -jar cnl-toolchain/target/cnl-toolchain-0.0.1-SNAPSHOT.jar`.
The rule OWL files, the mapping.txt and mapped.owl, and the result.owl files are created.
Currently, there are three issues:

1. The jar does not include all necessary resources. This does not break the execution.

```
0    [main] ERROR xt.validation.ResourceValidatorImpl  - The string resource '_UI_DiagnosticRoot_diagnostic' could not be located
java.util.MissingResourceException: The string resource '_UI_DiagnosticRoot_diagnostic' could not be located
```

2. The values are not correctly persisted
```
Exception in thread "main" com.complexible.stardog.StardogException: No value present
	at com.complexible.stardog.protocols.http.icv.client.ICVHttpConnection.executeAction(ICVHttpConnection.java:466)
	at com.complexible.stardog.protocols.http.icv.client.ICVHttpConnection.getViolationBindings(ICVHttpConnection.java:208)
	at com.complexible.stardog.icv.ICV.getProofs(ICV.java:259)
	at com.complexible.stardog.icv.api.AbstractICVConnection$1.proofs(AbstractICVConnection.java:90)
	at com.complexible.stardog.icv.api.AbstractICVConnection$1.proofs(AbstractICVConnection.java:87)
	at com.complexible.stardog.reasoning.AbstractStardogExplainer.proofs(AbstractStardogExplainer.java:80)
	at impl.StardogICVAPIImpl.explainViolationsForContext(StardogICVAPIImpl.java:182)
	at conformancecheck.impl.ConformanceCheckImpl.validateRule(ConformanceCheckImpl.java:69)
	at CNLToolchain.execute(CNLToolchain.java:182)
	at CNLToolchain.main(CNLToolchain.java:97)
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:141)
	at com.complexible.common.rdf.impl.StatementIterationInputStream.<init>(StatementIterationInputStream.java:106)
	at com.complexible.stardog.protocols.http.icv.client.ICVHttpConnection.executeAction(ICVHttpConnection.java:458)
	... 9 more
```
3. The paths need to the OWL files be adjusted. This includes the `projectPath` in CNLToolchain, the path to the `architectureconformance.owl` in `ConformanceCheckOntology`and the path to `famix.owl`and `main.owl` in `FamixOntologyTransformer`.

**Comment by nizap** [comment]
On the 1st "issue": See [this question](https://stackoverflow.com/q/38043360), it seems that solving this can be postponed.
On the 2nd "issue": The [StarDog API docs](https://www.stardog.com/docs/java/snarl/com/complexible/stardog/reasoning/stardogexplainer#proofs--) do not mention anything about the conditions under which this exception is thrown. Inspecting the created DB with StarDog Studio reveals that the tool chain has successfully written data to it. The output files written by the tool chain up to the exception look fine as well.

**Comment by vhschlenker** [comment]
With the commit c9e3daecbdd335d344e7d0debc35394f9739144a, the current state is as follows:

- The startog "No value present" exception is still present
- The created jar is usable as a cli tool with the following options:
```
Usage: ArchCNL CLI [-hV] [-c=<context>] [-d=<database>] [-s=<server>] <project 
                   path> <rules file>
      <project path>        The path to the project
      <rules file>          The path to the rules, relative to the project path
  -c, --context=<context>   Specifiy the context
  -d, --database=<database> Specifiy the database name
  -h, --help                Show this help message and exit.
  -s, --server=<server>     Specifiy the server
  -V, --version             Print version information and exit.
```

A sample usage after `mvn clean package` is therefore `java -jar cnl-toolchain/target/cnl-toolchain-0.0.1-SNAPSHOT.jar <<path to project>>/OnionArchitectureDemo/ architecture-documentation-onion.adoc`

**Comment by nizap** [comment]
The MissingResourceException has been fixed by b2d07f1f05886c754b5c00fb474d87edea49a402.

The StardogException appears (at least on my machine) only if the tool chain is executed outside of Eclipse (either using `java -jar <path-to-jar>` or `java -cp <path-to-jar> cnltoolchain.CNLToolchain`). However, it is not thrown when the tool chain is started from Eclipse (`cnl-toolchain -> Run As... -> Java Application`). All temporary files are written both successfully and correctly in either case. Also, all database operations up to the exception work as expected.  

My research on this problem has not identified any solution or more specific cause so far. My research included:

- searching for information online,
- printing any information about the variables used in the method where the error occurs (`StardogICVAPIImpl.explainViolationsForContext()`) and comparing the results between launch as a JAR and the launch from Eclipse (only unimportant differences like timestamps, size of the database, etc. have been observed),
- reading the Stardog API documentation of any method and class mentioned in the exception's stack trace,
- changing the Java version to 11, matching my installed JDK version,
- comparing the different `main()` methods used by the JAR and by the launch from Eclipse (no differences related to the problem),
- using the method `proof()` or `isValid()` instead of `proofs()` in `StardogICVAPIImpl.explainViolationsForContext()` (same exception is thrown),
- removing each of the "filtering" methods called on `aValidator` before `proofs()` is called (exception is still thrown),
- investigating the database state with Stardog Studio (only the information that would have been added after the exception is thrown is missing),
- changing  reasoning(false)`to `reasoning(true)` in the creation of `aConn` in `StardogICVAPIImpl.explainViolationsForContext()` (no effect).

**Comment by nizap** [comment]
Further experimentation revealed that the "Run As..." option in Eclipse (where the exception does not occur) includes the 527 JARs listed in the attached file. I suspect that at least one of them is incorrectly included into the "uber-JAR" produced by Maven Shade. I don't think that the JAR misses one of them, as Eclipse generated the list out of the `pom.xml` files. Moreover, building the JAR with `mvn clean package` displays several warnings regarding overlapping classes and resources between included JARs, further hinting to a problem in JAR inclusion.
[sorted-jars.txt](https://github.com/Mari-Wie/ArchCNL/files/5575199/sorted-jars.txt)

**Comment by nizap** [comment]
The exception was fixed by c96aa17641e7eb21b8cf634723ddef88137e1d07.
